local_planner/src/utils/trajectory_simulator.cpp

```diff
@@ -0,0 +1,107 @@
+#include "local_planner/trajectory_simulator.h"
+
+#include <cfloat>
+
+namespace {
+Eigen::Vector3f xy_norm_z_clamp(const Eigen::Vector3f& val, float max_xy_norm,
+                                float min_z, float max_z) {
+  Eigen::Vector3f result;
+  result.topRows<2>() = avoidance::norm_clamp<2>(val.topRows<2>(), max_xy_norm);
+  result.z() = std::min(max_z, std::max(min_z, val.z()));
+  return result;
+}
+
+float sqr(float f) { return f * f; }
+float cube(float f) { return f * f * f; }
+}
+
+namespace avoidance {
+
+TrajectorySimulator::TrajectorySimulator(
+    const avoidance::simulation_limits& config,
+    const avoidance::simulation_state& start, float step_time)
+    : config_(config), start_(start), step_time_(step_time) {}
+
+std::vector<simulation_state> TrajectorySimulator::generate_trajectory(
+    const Eigen::Vector3f& goal_direction, float simulation_duration) {
+  int num_steps = static_cast<int>(std::ceil(simulation_duration / step_time_));
+  std::vector<simulation_state> timepoints;
+  timepoints.reserve(num_steps);
+
+  const Eigen::Vector3f unit_goal = goal_direction.normalized();
+  const Eigen::Vector3f desired_velocity = xy_norm_z_clamp(
+      unit_goal * std::hypot(config_.max_xy_velocity_norm,
+                             unit_goal.z() > 0 ? config_.max_z_velocity
+                                               : config_.min_z_velocity),
+      config_.max_xy_velocity_norm, config_.min_z_velocity,
+      config_.max_z_velocity);
+
+  // calculate P and D constants such that they hit the jerk limit when
+  // doing accel from 0
+  float max_accel_norm = std::min(2 * std::sqrt(config_.max_jerk_norm),
+                                  config_.max_acceleration_norm);
+  float P_constant =
```

Comment:
This still bothers me a bit. It seems that P just needs to be something big and the relation between P and D has to be correct to be critically damped. But the actual value of P is not determined, this just sets a pseudo-scientific value for P. I wonder if there is an actual physically correct value, or if not if we should just put any high number instead.

Reply:
Agreed, let's look into this in the coming weeks.